File: Final/worker-python/worker.py
from flask import Flask, request, jsonify
import os
import time
import requests
import threading
import logging

app = Flask(__name__)

# Configuración de logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

# Inicialización de variables
task_queue = []
is_leader = False
leader_url = None
heartbeat_interval = 1  # Intervalo de tiempo en segundos para enviar heartbeats
heartbeat_failure_threshold = 3

worker_id = os.getenv('WORKER_ID', 'unknown')
worker_port = os.getenv('PORT', '5000')
hostname = os.getenv('HOSTNAME', 'localhost')
workers =  ["http://worker1:5001", "http://worker3:5003", "http://worker4:5004"]

heartbeat_failures = 0
current_worker_index = 0  # Índice del worker al que se asignará la próxima tarea

def process_task(task):
    logging.info(f"Processing task: {task['task_type']} for client {task['client_id']}")
    file_content = task['file_content']
    task_type = task['task_type']
    keyword = task.get('keyword')
    n = task.get('n')
    result = None
    if task_type == 'word_count':
        result = file_content.lower().count(keyword.lower()) if keyword else len(file_content.split())
        logging.info(f"Word count for client {task['client_id']} is {result}")
    elif task_type == 'keyword_search':
        result = keyword.lower() in file_content.lower()
        logging.info(f"Keyword search for client {task['client_id']} is {result}")
    elif task_type == 'keyword_repetition':
        result = file_content.lower().split().count(keyword.lower()) >= n
        logging.info(f"Keyword repetition for client {task['client_id']} is {result}")
    else:
        result = "Invalid task type"
        logging.info(f"Invalid task type for client {task['client_id']}")
    logging.info(f"Result for client {task['client_id']} is {result}")
    return result

@app.route('/submit_task', methods=['POST'])
def submit_task():
    try:
        task = request.json
        logging.info(f"Received task: {task}")
        result = process_task(task)
        return jsonify({"status": "Task processed successfully", "result": result}), 200
    except Exception as e:
        logging.error(f"Error processing task: {e}")
        return jsonify({"status": f"Error processing task: {e}"}), 400

@app.route('/heartbeat', methods=['POST'])
def heartbeat():
    global is_leader, leader_url, heartbeat_failures
    try:
        leader_id = request.json['worker_id']
        logging.info(f"Received heartbeat from worker {leader_id}")
        if leader_id != worker_id:
            is_leader = False
            leader_url = f"http://worker{leader_id}:{os.getenv('PORT')}"
        else:
            logging.info(f"Worker {worker_id} is the leader")
        heartbeat_failures = 0  # Reset failures on successful heartbeat
        return jsonify({"status": "Heartbeat received"}), 200
    except Exception as e:
        logging.error(f"Error receiving heartbeat: {e}")
        return jsonify({"status": f"Error receiving heartbeat: {e}"}), 400

@app.route('/is_leader', methods=['GET'])
def get_leader_status():
    logging.info(f"Checking leader status for worker {worker_id}")
    return jsonify({"is_leader": is_leader, "worker_id": worker_id}), 200

def send_heartbeat():
    global leader_url, is_leader, heartbeat_failures
    while True:
        time.sleep(heartbeat_interval)
        try:
            if is_leader:
                for worker in workers:
                    if worker != f"http://{hostname}:{worker_port}":
                        try:
                            response = requests.post(f"{worker}/heartbeat", json={"worker_id": worker_id})
                            logging.info(f"Heartbeat sent from leader {worker_id} to worker {worker}")
                        except Exception as e:
                            logging.error(f"Error sending heartbeat from leader {worker_id} to worker {worker}: {e}")
                            heartbeat_failures += 1
                            if heartbeat_failures >= heartbeat_failure_threshold:
                                logging.info(f"Leader {worker_id} failed. Transferring leadership.")
                                is_leader = False
                                transfer_leadership()
            else:
                logging.info(f"Worker {worker_id} is checking for leadership")
                check_for_leadership()
        except Exception as e:
            logging.error(f"Error in heartbeat management for worker {worker_id}: {e}")

def check_for_leadership():
    global is_leader, leader_url, heartbeat_failures
    for worker in workers:
        if worker != f"http://{hostname}:{worker_port}":
            try:
                response = requests.get(f"{worker}/is_leader")
                if response.status_code == 200:
                    data = response.json()
                    if data['is_leader']:
                        leader_url = worker
                        logging.info(f"Leader detected: {leader_url}")
                        return
            except Exception as e:
                logging.error(f"Error checking leader status from {worker}: {e}")
    is_leader = True
    leader_url = f"http://{hostname}:{worker_port}"
    logging.info(f"Worker {worker_id} assuming leadership")

def transfer_leadership():
    global is_leader, leader_url, heartbeat_failures
    heartbeat_failures = 0
    for worker in workers:
        if worker != f"http://{hostname}:{worker_port}":
            try:
                response = requests.get(f"{worker}/is_leader")
                if response.status_code == 200:
                    data = response.json()
                    if not data['is_leader']:
                        leader_url = worker
                        logging.info(f"New leader is {worker}")
                        return
            except Exception as e:
                logging.error(f"Error transferring leadership to {worker}: {e}")
    is_leader = True
    leader_url = f"http://{hostname}:{worker_port}"
    logging.info(f"Worker {worker_id} reassuming leadership due to lack of alternatives")

def assign_task(task):
    global current_worker_index
    attempts = 0
    while attempts < len(workers):
        worker = workers[current_worker_index]
        if worker != f"http://{hostname}:{worker_port}":  # No asignarse la tarea a sí mismo
            try:
                response = requests.post(f"{worker}/submit_task", json=task)
                if response.status_code == 200:
                    logging.info(f"Task assigned to {worker}")
                    current_worker_index = (current_worker_index + 1) % len(workers)
                    return response.json()
            except Exception as e:
                logging.error(f"Error assigning task to {worker}: {e}")
        current_worker_index = (current_worker_index + 1) % len(workers)
        attempts += 1
    return {"status": "Failed to assign task to any worker"}

@app.route('/leader_submit_task', methods=['POST'])
def leader_submit_task():
    if not is_leader:
        return jsonify({"status": "Not the leader"}), 403
    task = request.json
    logging.info(f"Leader {worker_id} received task: {task}")
    result = assign_task(task)
    return jsonify(result)

if __name__ == '__main__':
    if worker_id == '1':
        is_leader = True
        leader_url = f"http://{hostname}:{worker_port}"
        logging.info(f"Worker {worker_id} assuming leadership")
    threading.Thread(target=send_heartbeat).start()
    
    app.run(host='0.0.0.0', port=int(worker_port))

File: Final/worker-js/worker.js
const express = require('express');
const bodyParser = require('body-parser');
const axios = require('axios');

const app = express();
app.use(bodyParser.json());

const workerId = process.env.WORKER_ID || 'unknown';
const workerPort = process.env.PORT || 5000;
const hostname = process.env.HOSTNAME || 'localhost';
let workers =  ["http://worker1:5001", "http://worker3:5003", "http://worker4:5004"];

let isLeader = false;
let leaderUrl = null;
const heartbeatInterval = 1000; // Intervalo de tiempo en milisegundos para enviar heartbeats
const heartbeatFailureThreshold = 3;
let heartbeatFailures = 0;
let currentWorkerIndex = 0;  // Índice del worker al que se asignará la próxima tarea

const processTask = (task) => {
    console.log(`Processing task: ${task.task_type} for client ${task.client_id}`);
    const fileContent = task.file_content;
    const taskType = task.task_type;
    const keyword = task.keyword;
    const n = task.n;
    let result = null;

    if (taskType === 'word_count') {
        result = keyword ? (fileContent.match(new RegExp(keyword, 'gi')) || []).length : fileContent.split(/\s+/).length;
    } else if (taskType === 'keyword_search') {
        result = fileContent.toLowerCase().includes(keyword.toLowerCase());
    } else if (taskType === 'keyword_repetition') {
        result = (fileContent.match(new RegExp(keyword, 'gi')) || []).length >= n;
    } else {
        result = "Invalid task type";
    }

    console.log(`Result for client ${task.client_id} is ${result}`);
    return result;
};

app.post('/submit_task', (req, res) => {
    try {
        const task = req.body;
        console.log(`Received task: ${JSON.stringify(task)}`);
        const result = processTask(task);
        res.json({ status: "Task processed successfully", result });
    } catch (e) {
        console.error(`Error processing task: ${e}`);
        res.status(400).json({ status: `Error processing task: ${e}` });
    }
});

app.post('/heartbeat', (req, res) => {
    try {
        const leaderId = req.body.worker_id;
        console.log(`Received heartbeat from worker ${leaderId}`);
        if (leaderId !== workerId) {
            isLeader = false;
            leaderUrl = `http://${leaderId}:${workerPort}`;
        } else {
            console.log(`Worker ${workerId} is the leader`);
        }
        heartbeatFailures = 0;  // Reset failures on successful heartbeat
        res.json({ status: "Heartbeat received" });
    } catch (e) {
        console.error(`Error receiving heartbeat: ${e}`);
        res.status(400).json({ status: `Error receiving heartbeat: ${e}` });
    }
});

app.get('/is_leader', (req, res) => {
    console.log(`Checking leader status for worker ${workerId}`);
    res.json({ is_leader: isLeader, worker_id: workerId });
});

const sendHeartbeat = async () => {
    while (true) {
        await new Promise(resolve => setTimeout(resolve, heartbeatInterval));
        try {
            if (isLeader) {
                for (const worker of workers) {
                    if (worker !== `http://${hostname}:${workerPort}`) {
                        try {
                            await axios.post(`${worker}/heartbeat`, { worker_id: workerId });
                            console.log(`Heartbeat sent from leader ${workerId} to worker ${worker}`);
                        } catch (e) {
                            console.error(`Error sending heartbeat from leader ${workerId} to worker ${worker}: ${e}`);
                            heartbeatFailures += 1;
                            if (heartbeatFailures >= heartbeatFailureThreshold) {
                                console.log(`Leader ${workerId} failed. Transferring leadership.`);
                                isLeader = false;
                                transferLeadership();
                            }
                        }
                    }
                }
            } else {
                console.log(`Worker ${workerId} is checking for leadership`);
                checkForLeadership();
            }
        } catch (e) {
            console.error(`Error in heartbeat management for worker ${workerId}: ${e}`);
        }
    }
};

const checkForLeadership = async () => {
    for (const worker of workers) {
        if (worker !== `http://${hostname}:${workerPort}`) {
            try {
                const response = await axios.get(`${worker}/is_leader`);
                if (response.status === 200 && response.data.is_leader) {
                    leaderUrl = worker;
                    console.log(`Leader detected: ${leaderUrl}`);
                    return;
                }
            } catch (e) {
                console.error(`Error checking leader status from ${worker}: ${e}`);
            }
        }
    }
    isLeader = true;
    leaderUrl = `http://${hostname}:${workerPort}`;
    console.log(`Worker ${workerId} assuming leadership`);
};

const transferLeadership = async () => {
    heartbeatFailures = 0;
    for (const worker of workers) {
        if (worker !== `http://${hostname}:${workerPort}`) {
            try {
                const response = await axios.get(`${worker}/is_leader`);
                if (response.status === 200 && !response.data.is_leader) {
                    leaderUrl = worker;
                    console.log(`New leader is ${worker}`);
                    return;
                }
            } catch (e) {
                console.error(`Error transferring leadership to ${worker}: ${e}`);
            }
        }
    }
    isLeader = true;
    leaderUrl = `http://${hostname}:${workerPort}`;
    console.log(`Worker ${workerId} reassuming leadership due to lack of alternatives`);
};

const assignTask = async (task) => {
    let attempts = 0;
    while (attempts < workers.length) {
        const worker = workers[currentWorkerIndex];
        if (worker !== `http://${hostname}:${workerPort}`) {  // No asignarse la tarea a sí mismo
            try {
                const response = await axios.post(`${worker}/submit_task`, task);
                if (response.status === 200) {
                    console.log(`Task assigned to ${worker}`);
                    currentWorkerIndex = (currentWorkerIndex + 1) % workers.length;
                    return response.data;
                }
            } catch (e) {
                console.error(`Error assigning task to ${worker}: ${e}`);
            }
        }
        currentWorkerIndex = (currentWorkerIndex + 1) % workers.length;
        attempts += 1;
    }
    return { status: "Failed to assign task to any worker" };
};

app.post('/leader_submit_task', async (req, res) => {
    if (!isLeader) {
        return res.status(403).json({ status: "Not the leader" });
    }
    const task = req.body;
    console.log(`Leader ${workerId} received task: ${JSON.stringify(task)}`);
    const result = await assignTask(task);
    res.json(result);
});

if (workerId === '1') {
    isLeader = true;
    leaderUrl = `http://${hostname}:${workerPort}`;
    console.log(`Worker ${workerId} assuming leadership`);
}

app.listen(workerPort, () => {
    console.log(`Worker ${workerId} running on port ${workerPort}`);
    sendHeartbeat();
});
